cupy/random/_generator.py

```diff
@@ -1150,26 +1150,51 @@ def randint(self, low, high=None, size=None, dtype=int):
             - :func:`cupy.random.randint` for full documentation
             - :meth:`numpy.random.RandomState.randint`
         """
-        if high is None:
-            lo = 0
-            hi1 = int(low) - 1
+        if not (numpy.isscalar(low) and numpy.isscalar(high)):
+            low = cupy.asarray(low)
+            if high is None:
+                lo = cupy.zeros_like(low)
+                hi = low - 1
+            else:
+                lo = low
+                hi = cupy.asarray(high) - 1
+
+            if cupy.any(lo > high):
+                raise ValueError('low >= high')
+
+            if size is None:
+                size = cupy.broadcast(lo, hi).shape
+
+            diff = hi - lo
+            total_elems = functools.reduce(operator.mul, size, 1)
+            out = cupy.empty(total_elems, dtype=low.dtype)
+            for i, d in enumerate(cupy.flatiter(diff)):
+                out[i] = self._interval(d.item(), None)
```

Comment:
How do you think if it is possible to make `_interval` work as batch to draw samples with the upperbound of `diff` that may have different value across its elements?

Reply:
Let me take a look at the interval implementation

Reply:
@takagi, if the numbers are sampled as follow, then it is possible to sample numbers with different bounds (assuming `uint32` and no overflow cases):

```python
def _interval_nd(diff):
    mx = diff.flatten()
    mx1 = mx + 1
    sample = self._curand_generate(mx.size, np.uint32)
    sample %= mx1
    return sample
```

Reply:
I have some questions regarding the overflow management. 

1. In the current implementation the following max bound (on `uint64`) is imposed: https://github.com/cupy/cupy/blob/d588608a45e14d41bda6057cdb23daef3e40ae85/cupy/random/_generator.py#L676 This works fine on Python integers, since they operate over bigint arithmetic, however, since ndarrays operate on fixed-precision numbers, then it couldn't be possible to store `1 << 64` (the total `uint64` numbers represented) on a ndarray, which would prevent the computation of the `upper_bound` on `(u)int64` inputs.
2. Since `mx1` would be an fixed-precision ndarray, then would be this check redundant since the output numbers will be on a fixed precision? https://github.com/cupy/cupy/blob/d588608a45e14d41bda6057cdb23daef3e40ae85/cupy/random/_generator.py#L687

Reply:
Thanks for your points.

1. Yes, but when `mx < _UINT64_MAX`, we can use `(_UINT64_MAX - mx) % (mx + 1)` equivalent to `(1 << 64) % (mx + 1)`, which we can compute in uint64. When `mx == _UINT64_MAX`, `(1 << 64) % (mx + 1)` is clearly zero.
2. Also yes, but let's keep the branch as a fast path in case mx is a scalar. Masking is much performant when possible.